src/error.rs

```diff
@@ -1,141 +1,351 @@
+//! Module containing error details.
+
 /// An error that can occur during variable substitution.
 #[derive(Debug, Clone)]
 #[cfg_attr(test, derive(Eq, PartialEq))]
-pub struct Error {
-	/// The private error details.
-	inner: ErrorInner,
+pub enum Error {
```

Comment:
Do you feel something like `#[non_exhaustive]`, or better yet, something like `Custom(String)`/`Custom(Box<dyn Error + Send>)` is warranted here?

I feel ambivalent myself. While I'm definitely worried this is exposing too much implementation details, it's also not clear whether future expansions are actually necessary in this case.

Reply:
Right, plus, my feeling is: if someone is manually handling error cases, then breaking compilation when we add a new error case is a good thing.

Or maybe a compiler warning would be good enough, but that is not really possible.

Reply:
Of course, if you only handle *one* case specifically, we won't break compilation by adding a new case. But we will have to do a major bump in case someone is handling all cases.